python/ray/data/grouped_dataset.py

```diff
@@ -258,22 +260,24 @@ def get_key(row):
                 return None
 
         # Returns the group boundaries.
-        def get_boundaries(block):
+        def get_key_boundaries(batch):
             boundaries = []
-            pre = None
-            for i, item in enumerate(block.iter_rows()):
-                if pre is not None and get_key(pre) != get_key(item):
-                    boundaries.append(i)
-                pre = item
-            if block.num_rows() > 0:
-                boundaries.append(block.num_rows())
+            block_accessor = BlockAccessor.for_block(batch)
+            keys = block_accessor.get_keys(self._key)
```

Comment:
Probably it's cleaner to handle self._key is None case here so `get_keys` can assume the key is never None.

Reply:
This is passing in a KeyFn-typed parameter to get_keys, so get_keys() is already in position to handle it anyway since KeyFn can be None.